abdm_integrator/hip/views/care_contexts.py

```diff
@@ -0,0 +1,108 @@
+        serializer = LinkCareContextRequestSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        self.ensure_not_already_linked(serializer.data)
+        gateway_request_id = self.gateway_add_care_contexts(serializer.data)
+        self.save_link_request(request.user, gateway_request_id, serializer.data)
+        response_data = poll_and_pop_data_from_cache(gateway_request_id)
+        return self.generate_response_from_callback(response_data)
+
+    def ensure_not_already_linked(self, request_data):
+        care_contexts_references = [care_context['referenceNumber']
+                                    for care_context in request_data['patient']['careContexts']]
+        linked_care_contexts = list(LinkCareContext.objects.filter(
+            reference__in=care_contexts_references,
+            link_request_details__hip_id=request_data['hip_id'],
+            link_request_details__patient_reference=request_data['patient']['referenceNumber'],
+            link_request_details__status=LinkRequestStatus.SUCCESS
+        ).values_list('reference', flat=True))
+
+        if linked_care_contexts:
+            code = HIPError.CODE_CARE_CONTEXT_ALREADY_LINKED
+            message = HIPError.CUSTOM_ERRORS[code].format(linked_care_contexts)
+            raise CustomError(
+                error_code=code,
+                error_message=message,
+            )
+
+    def gateway_add_care_contexts(self, request_data):
+        payload = ABDMRequestHelper.common_request_data()
+        payload['link'] = deepcopy(request_data)
+        ABDMRequestHelper().gateway_post(HIPGatewayAPIPath.ADD_CARE_CONTEXTS, payload)
+        return payload['requestId']
+
+    @transaction.atomic()
+    def save_link_request(self, user, gateway_request_id, request_data):
+        link_request_details = LinkRequestDetails.objects.create(
+            hip_id=request_data['hip_id'],
+            patient_reference=request_data['patient']['referenceNumber'],
+            patient_display=request_data['patient']['display'],
+            initiated_by=LinkRequestInitiator.HIP
+        )
+        HIPLinkRequest.objects.create(user=user, gateway_request_id=gateway_request_id,
+                                      link_request_details=link_request_details)
+        # Store any additional info related to care context (* Do not store health data)
+        additional_info = {'domain': getattr(user, 'domain', None)}
+        link_care_contexts = [LinkCareContext(reference=care_context['referenceNumber'],
+                                              display=care_context['display'],
+                                              health_info_types=care_context['hiTypes'],
+                                              additional_info=additional_info,
+                                              link_request_details=link_request_details)
+                              for care_context in request_data['patient']['careContexts']]
```

Comment:
I see that you are used to this style, but in all of HQ we use a different style. Consider using the same style for consistency.
```suggestion
        link_care_contexts = [
                LinkCareContext(
                        reference=care_context['referenceNumber'],
                        display=care_context['display'],
                        health_info_types=care_context['hiTypes'],
                        additional_info=additional_info,
                        link_request_details=link_request_details
                )
                for care_context in request_data['patient']['careContexts']
        ]
```

Reply:
Addressed in 5007f02ee2e12897e8ee4b8a551266b4b0bfb0f1